File: 2023/go/pkg/day4/problem1.go
package day4

import (
	"fmt"
	"slices"
	"strings"
)

func SolveProblem1(lines []string) string {
	sum := 0

	for _, line := range lines {
		cards := strings.Split(line, ": ")
		allnums := strings.Split(cards[1], " | ")

		winning := strings.Fields(allnums[0])
		mynums := strings.Fields(allnums[1])

		points := 0

		for _, num := range mynums {
			if slices.Contains(winning, num) {
				if points == 0 {
					points = 1
				} else {
					points *= 2
				}
			}
		}

		sum += points
	}

	return fmt.Sprint(sum)
}

File: 2023/rust/src/day4/problem1.rs
pub fn solve(lines: Vec<&str>) -> String {
    let mut sum = 0;

    for line in lines.iter() {
        let card: Vec<&str> = line.split(": ").collect();
        let allnums: Vec<&str> = card[1].split(" | ").collect();

        let winning: Vec<&str> = allnums[0].split_whitespace().collect();

        let mut points = 0;
        for num in allnums[1].split_whitespace() {
            if winning.contains(&num) {
                if points == 0 {
                    points = 1;
                } else {
                    points *= 2;
                }
            }
        }

        sum += points;
    }

    sum.to_string()
}
